src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private Object[] data;
+
+    public ArrayList() {
+        data = new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == data.length) {
+            grow();
+        }
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index <= size && index >= 0) {
+            Object[] buffer = new Object[size - index];
+            System.arraycopy(data, index, buffer, 0, size() - index);
+            size -= size - index;
+            if (size + buffer.length > data.length) {
+                grow();
+            }
+            data[index] = value;
+            System.arraycopy(buffer, 0, data, index + 1, buffer.length);
+            size += buffer.length + 1;
+        } else {
+            throw new ArrayIndexOutOfBoundsException();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (size + list.size() > data.length) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            data[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
```

Comment:
the index check for `get()`, `set()` and `remove()` methods may be moved to the private method
also, what if index is negative?